include/onnxruntime/core/optimizer/rewrite_rule.h

```diff
@@ -49,7 +51,11 @@ class RewriteRule {
     return desc_;
   }
 
-  /** Checks if the condition of the rule is satisfied, and if so applies the rule.
+  const std::unordered_set<std::string>& TargetOpTypes() const noexcept {
```

Comment:
Is this going to be used as a set? Right now, a vector would do based on your usage. The code to build the map is large. Also, the way the RewriteRule constructor is called, the compiler is generating a local unordered_set (or vector) on the stack and then that is copied into the base class field and the original is deleted. Maybe just pass down an initializer_list instead and not inline RewriteRule?

Reply:
I used a set given there should be no duplicates. But a vector would be fine.
I will also take care of the rest.

Reply:
As we discussed, I instead made the TargetOpTypes virtual, so that we avoid having the rewrite rules keeping a copy of the list for the whole life of the object.
I kept the set to make sure nobody adds duplicates by mistake, but I can change it to vector.
Re: the code to build the map being large, I used a for_each with a lambda instead, but did not make the code smaller. Did you have something else in mind to make it more succinct?